File: packages/nhost-swift/Sources/Nhost/Auth/PKCE.swift
import Foundation

/// PKCE (Proof Key for Code Exchange) helpers for RFC 7636, mirroring nhost-js's
/// `generateCodeVerifier` / `generateCodeChallenge` / `generatePKCEPair`. Output is
/// byte-compatible with the JavaScript SDK: base64url without padding.
public enum PKCE {
    /// Generates a cryptographically random code verifier (43 base64url characters).
    public static func generateCodeVerifier() -> String {
        var generator = SystemRandomNumberGenerator()
        let bytes = (0..<32).map { _ in UInt8.random(in: .min ... .max, using: &generator) }

        return base64url(Data(bytes))
    }

    /// Derives the S256 code challenge from a code verifier.
    public static func generateCodeChallenge(from verifier: String) -> String {
        base64url(NhostSHA256.hash(Data(verifier.utf8)))
    }

    /// Generates a code verifier and its S256 challenge in one call.
    public static func generatePair() -> (verifier: String, challenge: String) {
        let verifier = generateCodeVerifier()

        return (verifier: verifier, challenge: generateCodeChallenge(from: verifier))
    }

    static func base64url(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}

File: packages/nhost-rust/src/auth/pkce.rs
//! PKCE (Proof Key for Code Exchange) utilities for RFC 7636.

use base64::Engine;
use rand::RngCore;
use sha2::{Digest, Sha256};

fn b64url(data: &[u8]) -> String {
    base64::engine::general_purpose::URL_SAFE_NO_PAD.encode(data)
}

/// Generates a cryptographically random PKCE code verifier (43 base64url
/// characters, per RFC 7636).
pub fn generate_code_verifier() -> String {
    let mut buf = [0u8; 32];
    rand::rng().fill_bytes(&mut buf);
    b64url(&buf)
}

/// Derives an S256 code challenge from a code verifier.
pub fn generate_code_challenge(verifier: &str) -> String {
    let digest = Sha256::digest(verifier.as_bytes());
    b64url(&digest)
}

/// A PKCE code verifier and its derived S256 challenge.
#[derive(Clone)]
pub struct PkcePair {
    /// The high-entropy secret sent only when exchanging the authorization code.
    pub verifier: String,
    /// The public S256 digest sent with the initial authorization request.
    pub challenge: String,
}

impl std::fmt::Debug for PkcePair {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("PkcePair")
            .field("verifier", &"<redacted>")
            .field("challenge", &self.challenge)
            .finish()
    }
}

/// Generates a PKCE code verifier and its S256 challenge.
pub fn generate_pkce_pair() -> PkcePair {
    let verifier = generate_code_verifier();
    let challenge = generate_code_challenge(&verifier);
    PkcePair {
        verifier,
        challenge,
    }
}
